### ISSUE TITLE
Disable the extension when Tracking Protection is set to "always"

### ISSUE THREAD

**Issue by groovecoder** [LaserlikeInc/firefox-addon#60]
Please disable the extension's data collection from all windows if `browser.privacy.websites.trackingProtectionMode` returns `always`. There should be no browser history tracking when the user has enabled tracking protection.

I'm not sure if it's possible in Test Pilot, but to improve the UX we should make it clear to users that this experiment is incompatible with Tracking Protection.

**Comment by sriniven** [comment]
Fixed, but please see this explanation.
We have backend calls from the sidebar, as well as a background task that is common to all windows/tabs.
- When 'do not track' is on, we do not call our backend from either the sidebar or the background task, and do not display any results in the sidebar. 
- For private window, we do not call backend from the sidebar. However, the 'background task' still calls our backend: this does not have access to the user cookie, it hits a public URL for the Read Next. 
This is because we do not know if the call is from a private window in the background task. This call is not associated with the user, so we won't track the user.
If there is a way to tell if the URL is coming from the private window, we can disable this backend call from the background task.

**Comment by chuckharmston** [comment]
Thanks Cheenu! Can you also please make sure to check against `trackingProtectionMode`, as described in this bug and an email from me earlier today? This is similar but separate to Do Not Track.

**Comment by sriniven** [comment]
hi Chuck,

We are doing this in the code as suggested in the bug.
However, please note the explanation about the background task. In the
background task,
which is common to all windows and tabs, this check is not preventing the
backend call.
However this backend call is 'anonymous', it has no user information.
To avoid this call, we need a API that can return the
trackingProtectionMode of a window given the window id.
Then in the background task we can check that the window from the URL came
is private and not hit the backend.

if (browser.privacy.websites.trackingProtectionMode === ‘always’ ||
   browser.privacy.websites.trackingProtectionMode.ALWAYS === ‘always’ ||
   navigator.doNotTrack === ‘1’) {
   Log(‘do not track enabled, return’);
   return;
 }

thanks,
Cheenu

**Comment by chuckharmston** [comment]
<img width="1629" alt="tp" src="https://user-images.githubusercontent.com/23885/43748246-c848f5a6-99ab-11e8-9af3-531171bf179e.png">

<img width="1597" alt="dnt" src="https://user-images.githubusercontent.com/23885/43748249-c9c21d4a-99ab-11e8-8978-32d801e2be60.png">

In my testing, version 0.9.35230 of the add-on no longer shows UI or makes any requests to Laserlike when either DNT or TP are enabled.

Ideally we'd see the "While browsing privately..." error message in these cases, but I think this is enough to unblock.

I'll leave this open for QA to verify in the morning.

**Comment by sriniven** [comment]
Thanks, Chuck. Note that in the private browsing mode, if you go to Network tab, you will still see the anonymous backend calls, but the sidebar won't show anything.

We'll add the "While browsing privately.." kind of UI tomorrow, but would be good to unblock without that like you have said.

**Comment by groovecoder** [comment]
Lack of a user-identifying cookie has traditionally NOT been considered anonymous (enough) for Private Browsing. 

The laserlike server will still receive an IP address, and can potentially receive sensitive PII via the kinds of full urls that are sent in the `Referer`-like `search_like` `target` values.

E.g., I used Private Browsing to visit healthcare.gov to shop for a health care provider under a pseudonym. Here's a sequence of `target` values that were sent to laserlike.com:

```
u:https://idahohix.yourhealthidaho.org/hix/preeligibility
u:https://idahohix.yourhealthidaho.org/hix/private/preferences
u:https://idahohix.yourhealthidaho.org/hix/private/planselection?insuranceType=HEALTH
u:https://www.rxeob.com/patientdashboard_sh/advanced_drug_lookup.aspx
u:https://www.rxeob.com/patientdashboard_sh/drug_lookup_results.aspx?ndc=00009002902&ln=XANAX%200.25%20MG%20TABLET
u:https://sbmsso.idalink.idaho.gov/authenticationendpoint/login.do?RelayState=0&SAMLRequest=fZFRb4IwFIXf9ytI36WAU7QRjJsxmriMCC7Z3ipWWget6y3E%2FftV0MS9mPSlt%2Fd8PTlnMj1XpdMwDULJCPmuhxwmc7UXsojQNlv0RmgaP02AVmVwIrPacLlhPzUD48wAmDZW96ok1BXTKdONyNl2s44QN%2BYEBGOxp1xxcXZ%2FVa05o6Xh7chVusB2ji9knKbvmJaCAv5crpAzt3ghqWk93UiwqwCUa9WlkN9uRylU0xLsC3IWSuestRgho2uGnNU8QjQU%2FLA7Dnh4fBZB3heFf%2FT6ey8cDpk%2FtjuQUADRsAgdaAkXFUDNVhIMlSZCgeePep49Yeb5ZDAk3sgd9cdfyEm0MipX5YuQXV61lkRREEAkrRgQk5N09rYmgeuRXbcEZJllSS95TzPkfNxyDy652yYkkC7px6zT9WMUd8WQ1rG%2BJzwG0Ft1KLaBT%2FA9Jb5e%2F7cd%2FwE%3D&SigAlg=http%3A%2F%2Fwww.w3.org%2F2000%2F09%2Fxmldsig%23rsa-sha1&Signature=DF%2BwScn7dzUlXHG2aHJZY1c2CjaaOiFHkN97eE7LgQtTJkzq5G84oRm5ma%2FJSpJb3c1W08Fq1IeYOzFJx0r%2BKuGy4EIGGn6%2F5rUcW%2FbtYt%2BgCEWzqrGJ4uo8pDyC5yIZNUEz%2B5Afm5QTd4x%2FD1nI2zYybAp2gR02eAR35dyd3uMVBEW%2FAwgkg6lpOfdJz8haSuJ05ti%2Bnws%2BjHIIvMAz6n%2BZzcNU60htPVwZR%2B89qtlosF%2Fh9KkBZzWP0zt5pLWR87qNGHjdhin7gBk%2FqujvzonGmSfU8Nwm2Ed6V%2FdEI70HVHANG9ZpKHjJPS98aEvJPA%2B3ARpMD6hm%2B412D%2B4haQ%3D%3D&commonAuthCallerPath=%252Fsamlsso&forceAuth=true&passiveAuth=false&relyingParty=YHI&tenantDomain=carbon.super&type=samlsso&sessionDataKey=3e11f6ab-0393-4fd6-b19d-bb630d5ff3df&relyingParty=YHI&type=samlsso&sp=YHI&isSaaSApp=false&authenticators=BasicAuthenticator:LOCAL
u:https://sbmsso.idalink.idaho.gov/samlsso?sessionDataKey=955e9c34-a8da-4deb-abf2-b17492fd2748
u:https://www.yourhealthidaho.org/?s=spiny_norman8221%40sudomail.com&submit=Go
```

So in addition to passive IP address collection, this includes state, drug name, SAML data, and a personal email address in the `search_like` `target` values.

**Comment by sriniven** [comment]
We are adding what Chuck suggested, will update as soon as it is done.

**Comment by SoftVision-CosminMuntean** [comment]
I have verified this issue using latest Laserlike (0.9.41) version and the "While browsing privately or with tracking protection..." message is correctly displayed in the sidebar if "Tracking Protection" is set to "Always". Also, I haven't observed any Laserlike requests on Network tab.
Tested on Windows 7 x64, Mac 10.13 and Arch Linux.

Considering this I will close this issue as verified fixed.
